sys/vm/vm_fault.c

```diff
@@ -1486,7 +1466,7 @@ vm_fault(vm_map_t map, vm_offset_t vaddr, vm_prot_t fault_type,
 				faultcount = 1;
 
 		} else {
-			fs.prot &= ~VM_PROT_WRITE;
+			fs.prot &= ~(VM_PROT_WRITE | VM_PROT_WRITE_CAP);
```

Comment:
```suggestion
			fs.prot &= ~(VM_PROT_WRITE | VM_PROT_WRITE_CAP);
			if ((fs.object & OBJ_HASCAP) == 0)
				fs.prot &= ~VM_PROT_READ_CAP;
```

Reply:
You'd have to make a change like this (and one other place) to avoid setting READ_CAP on the pre-COW mapping.  I think pmap_enter_quick() and pmap_enter_object() might also need to have their `prot` massaged explicitly to clear READ_CAP perhaps.